Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,112 +61,155 @@ bool VirtIOConsole::handle_device_config_change()
 
 void VirtIOConsole::handle_queue_update(u16 queue_index)
 {
-    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update");
-    VERIFY(queue_index <= TRANSMITQ);
-    switch (queue_index) {
-    case RECEIVEQ: {
-        auto& queue = get_queue(RECEIVEQ);
+    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update {}", queue_index);
+
+    if (queue_index == CONTROL_RECEIVEQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_receive_buffer->lock());
+        auto& queue = get_queue(CONTROL_RECEIVEQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
 
-        ScopedSpinLock ringbuffer_lock(m_receive_buffer->lock());
-
-        auto used_space = m_receive_buffer->reserve_space(used).value();
-        auto remaining_space = RINGBUFFER_SIZE - used;
-
-        // Our algorithm always has only one buffer in the queue.
-        VERIFY(!queue.new_data_available());
-        popped_chain.release_buffer_slots_to_queue();
+        while (!popped_chain.is_empty()) {
+            popped_chain.for_each([&](auto addr, auto) {
+                auto offset = addr.as_ptr() - m_control_receive_buffer->start_of_region().as_ptr();
+                ControlMessage* message = (ControlMessage*)m_control_receive_buffer->vaddr().offset(offset).as_ptr();
```

Comment:
No need to state this type twice, `auto message = (ControlMessage*)...;` already contains the type.